protocol/ntpshm/ntpshm.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright (c) Facebook, Inc. and its affiliates.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ntpshm
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// SHMKEY is a key of the first SHM segment
+const SHMKEY = 0x4e545030
+
+// IPC_CREAT create if key is nonexistent
+const IPC_CREAT = 00001000
+
+// IPC_EXCL fail if key exists
+const IPC_EXCL = 00002000
+
+/* NTPSHM Declaration of the SHM segment from ntp (ntpd/refclock_shm.c) */
+type NTPSHM struct {
+	mode                 int32
+	count                int32
+	clockTimeStampSec    uint
+	clockTimeStampUSec   int32
+	receiveTimeStampSec  uint
+	receiveTimeStampUSec int32
+	leap                 int32
+	precision            int32
+	nsamples             int32
+	valid                int32
+	clockTimeStampNSec   uint32
+	receiveTimeStampNSec uint32
+	dummy                [8]uint32
+}
+
+// Create a segment in SHM and return the ID
+func Create() (uintptr, error) {
+	shmID, _, errno := unix.Syscall(unix.SYS_SHMGET, uintptr(SHMKEY), 0, uintptr(IPC_CREAT|IPC_EXCL|0600))
+	if errno != 0 {
+		return 0, fmt.Errorf("failed get shm: %d", errno)
+	}
+	return shmID, nil
+}
+
+// ReadID reads SHM segment by ID
+func ReadID(id uintptr) (*NTPSHM, error) {
+	shmptr, _, errno := unix.Syscall(unix.SYS_SHMAT, id, 0, 0)
+	if errno != 0 {
+		return nil, fmt.Errorf("failed to attach to shm: %d", errno)
+	}
+
+	var shm []NTPSHM
+	sh := (*reflect.SliceHeader)(unsafe.Pointer(&shm))
+	sh.Data = shmptr
+	sh.Len = 1
+	sh.Cap = 1
+
+	return &shm[0], nil
+}
+
+// Read SHM segment
+func Read() (*NTPSHM, error) {
+	shmID, _, errno := unix.Syscall(unix.SYS_SHMGET, uintptr(SHMKEY), 0, uintptr(0400))
```

Comment:
what does `0400` mean and where does it come from?

Reply:
It's a standard Unix permissions. This is to open shared memory segment for reading.